# Logistic Regression

## Part 1 - Data Preprocessing

### Importing the dataset

In [126]:
# Step 1: Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [127]:
import pandas as pd
dataset = pd.read_csv('adult.csv')

In [128]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             30161 non-null  int64
 1   workclass       30161 non-null  int64
 2   fnlwgt          30161 non-null  int64
 3   education       30161 non-null  int64
 4   education.num   30161 non-null  int64
 5   marital.status  30161 non-null  int64
 6   occupation      30161 non-null  int64
 7   relationship    30161 non-null  int64
 8   race            30161 non-null  int64
 9   capital.gain    30161 non-null  int64
 10  capital.loss    30161 non-null  int64
 11  hours.per.week  30161 non-null  int64
 12  native.country  30161 non-null  int64
 13  income          30161 non-null  int64
 14  sex             30161 non-null  int64
dtypes: int64(15)
memory usage: 3.5 MB


In [129]:
dataset.info

<bound method DataFrame.info of        age  workclass  fnlwgt  education  education.num  marital.status  \
0       34         13   67083         20             13              40   
1       27         10  292472         21             10              40   
2       40         10  121956         20             13              41   
3       32         10  112137         22              1              42   
4       37         10  191342         23              2              42   
...    ...        ...     ...        ...            ...             ...   
30156   43         10  557349         34              6              42   
30157   22         10  180060         25              9              40   
30158   29         10  273051         20             13              42   
30159   36         11  337778         21             10              42   
30160   41         12  145441         29             11              42   

       occupation  relationship  race  capital.gain  capital.loss  

In [130]:
dataset.head(10)

age  workclass  fnlwgt  education  education.num  marital.status  \
0   34         13   67083         20             13              40   
1   27         10  292472         21             10              40   
2   40         10  121956         20             13              41   
3   32         10  112137         22              1              42   
4   37         10  191342         23              2              42   
5   25         10  228608         21             10              40   
6   27         10  292472         21             10              42   
7   40         10  173858         24              4              42   
8   51         10   63685         25              9              42   
9   28         10  149769         25              9              40   

   occupation  relationship  race  capital.gain  capital.loss  hours.per.week  \
0          50            70    80          1471             0              40   
1          51            71    80             0          1876              45   
2          52            71    80         13550             0              40   
3          53            72    80          4508             0              40   
4          51            73    80             0             0              40   
5          51            74    80             0             0              40   
6          52            73    80             0             0              40   
7          54            73    80             0             0              42   
8          55            73    80             0             0              50   
9          53            71    80             0             0              40   

   native.country  income  sex  
0              90       5    1  
1              90       5    1  
2              90       6    1  
3              90       5    0  
4              90       5    1  
5              90       5    0  
6              90       6    1  
7              90       5    1  
8              90       5    1  
9              90       5    1

### Getting the Inputs and Outputs

In [184]:
#row,column
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values

In [166]:
X

array([[    34,     13,  67083, ...,     40,     90,      5],
       [    27,     10, 292472, ...,     45,     90,      5],
       [    40,     10, 121956, ...,     40,     90,      6],
       ...,
       [    29,     10, 273051, ...,     52,    130,      6],
       [    36,     11, 337778, ...,     60,    130,      6],
       [    41,     12, 145441, ...,     40,    130,      5]])

In [186]:
y

array([1, 1, 1, ..., 1, 1, 1])

### Creating the Training Set and Test Set|

In [187]:
from sklearn.model_selection import train_test_split #library, module, function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [188]:
X_train

array([[    38,     12, 344480, ...,     40,    128,      5],
       [    55,     10, 143372, ...,     40,    128,      5],
       [    39,     12,  29874, ...,     60,    128,      6],
       ...,
       [    18,     10, 107277, ...,     20,    128,      5],
       [    29,     10, 632593, ...,     40,    128,      5],
       [    41,     13, 133060, ...,     40,    128,      6]])

In [189]:
X_test

array([[    19,     10, 128363, ...,     30,    128,      5],
       [    21,     10, 191444, ...,     40,    128,      5],
       [    37,     10, 214738, ...,     40,    128,      5],
       ...,
       [    22,     10, 445758, ...,     40,    115,      5],
       [    19,     10,  60890, ...,     49,    128,      5],
       [    30,     14, 295612, ...,     40,    128,      5]])

In [190]:
y_train

array([1, 0, 1, ..., 0, 1, 1])

In [191]:
y_test

array([0, 1, 1, ..., 1, 1, 0])

### Feature Scaling

In [192]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)

In [193]:
X_train

array([[-0.03908752,  0.76150077,  1.47511931, ..., -0.07622301,
         0.26531338, -0.57696741],
       [ 1.25137968, -0.52497806, -0.44067986, ..., -0.07622301,
         0.26531338, -0.57696741],
       [ 0.03682231,  0.76150077, -1.52188684, ...,  1.59182732,
         0.26531338,  1.73320014],
       ...,
       [-1.55728424, -0.52497806, -0.78452879, ..., -1.74427335,
         0.26531338, -0.57696741],
       [-0.72227605, -0.52497806,  4.21974731, ..., -0.07622301,
         0.26531338, -0.57696741],
       [ 0.18864198,  1.40474019, -0.53891425, ..., -0.07622301,
         0.26531338,  1.73320014]])

### Part 2 - Building and Training the Model

### Building the Model

In [194]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(random_state=0)

### Training the Model

In [195]:
model.fit(X_train,y_train)

LogisticRegression(random_state=0)

### Inference

#### Making the predictons of the data points in the test set

In [196]:
y_pred = model.predict(sc.transform(X_test))

In [197]:
y_pred

array([1, 0, 1, ..., 1, 1, 0])

In [198]:
y_test

array([0, 1, 1, ..., 1, 1, 0])

## Making the Prediction of A Single Data Point with:

### work class
1. Private 10
2. Self emp inc 11
3. Self-emp-not-inc 12
4. Federal-gov 13
5. State-gov 14
6. Local-gov 15
7. Without-pay 16
8. Never-worked 17

## education
1. Bachelors 20
2. Some-college 21
3. Preschool 22
4. 1st-4th 23
5. 7th-8th 24
6. HS-grad 25
7. 12th 26
8. Prof-school 27
9. 11th 28
10. Assoc-voc 29
11. Assoc-acdm 30
12. Masters 31
13. Doctorate 32
14. 9th 33
15. 10th 34
16. 5th-6th 35

## marital.status
1. Never-married 40
2. Married-spouse-absent 41
3. Married-civ-spouse 42
4. Divorced 43
5. Widowed 44
6. Separated 45
7. Married-AF-spouse 46

## occupation
1. Exec-managerial 50
2. Craft-repair 51
3. Prof-specialty 52
4. Machine-op-inspct 53
5. Other-service 54
6. Sales 55
7. Farming-fishing 56
8. Adm-clerical 57
9. Transport-moving 58
10. Tech-support 59
11. Handlers-cleaners 60
12. Protective-serv 61
13. Priv-house-serv 62
14. Armed-Forces 63

## relationship
1. Unmarried 70
2. Not-in-family 71
3. Wife 72
4. Husband 73
5. Other-relative 74
6. Own-child 75

## race
1. Asian-Pac-Islander 80
2. White 81
3. Other 82
4. Amer-Indian-Eskimo 83
5. Black 84

## sex 
1. Male 1
2. Female 0

## native.country
1. Cambodia 90
2. Canada 91
3. China 92
4. Columbia 93
5. Cuba 94
6. Dominican-Republic 95
7. Ecuador 96
8. El-Salvador 97
9. England 98
10. France 99
11. Germany 100
12. Greece 101
13. Guatemala 102
14. Haiti 103
15. Holand-Netherlands 104
16. Honduras 105
17. Hong 106
18. Hungary 107 
19. India 108
20. Iran 109
21. Ireland 110
22. Italy 111
23. Jamaica 112
24. Japan 113
25. Laos 114
26. Mexico 115
27. Nicaragua 116
28. Outlying-US(Guam-USVI-etc) 117
29. Peru 118
30. Philippines 119
31. Poland 120
32. Portugal 121
33. Puerto-Rico 122
34. Scotland 123
35. South 124
36. Taiwan 125
37. Thailand 126
38. Trinadad&Tobago 127
39. United-States 128
40. Vietnam 129
41. Yugoslavia 130


age	workclass	fnlwgt	education	education.num	marital.status	occupation	relationship	race	capital.gain	capital.loss	hours.per.week	native.country	income


In [199]:
model.predict(sc.transform([[34,	13,	67083,	20,	13,	40,	50,	70,	80,	1471,	0,	40,	90,	5]]))

array([0])

In [201]:
model.predict(sc.transform([[27,	10,	292472,	21,	10,	40,	51,	71,	80,	0,	1876,	45,	90,	5]]))

array([1])

## Part 3. Evaluating the Model

### Confusion Matrix

In [202]:
from sklearn.metrics import confusion_matrix
confusion_matrix (y_test, y_pred)

array([[ 884, 1088],
       [ 525, 3536]])

### Accuracy

In [203]:
( 884+3536)/( 884+525+1088+3536)

0.7326371622741588

In [204]:
from sklearn.metrics import accuracy_score
accuracy_score (y_test, y_pred)

0.7326371622741588